src/Xamarin.Android.Build.Tasks/Tasks/GenerateJavaStubs.cs

```diff
@@ -285,13 +290,62 @@ void WriteTypeMappings (List<TypeDefinition> types)
 			TypeNameMapGenerator createTypeMapGenerator () => UseSharedRuntime ?
 				new TypeNameMapGenerator (types, logger) :
 				new TypeNameMapGenerator (ResolvedAssemblies.Select (p => p.ItemSpec), logger);
-			using (var gen = createTypeMapGenerator ())
-			using (var stream = new MemoryStream ()) {
-				gen.WriteJavaToManaged (stream);
-				MonoAndroidHelper.CopyIfStreamChanged (stream, Path.Combine (OutputDirectory, "typemap.jm"));
-				stream.SetLength (0); //Reuse the stream
-				gen.WriteManagedToJava (stream);
-				MonoAndroidHelper.CopyIfStreamChanged (stream, Path.Combine (OutputDirectory, "typemap.mj"));
+			using (var gen = createTypeMapGenerator ()) {
+				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.jm"), "jm", gen.WriteJavaToManaged);
+				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.mj"), "mj", gen.WriteManagedToJava);
+			}
+		}
+
+		void UpdateWhenChanged (string path, string type, Action<Stream> generator)
```

Comment:
This looks like it partially reverted a perf improvement from: 9b99cce4fac384a6bea4234b88d729b696022fd9

Can we get rid of the `UpdateWhenChanged` method and put your new code inside the `using (var gen ...)` block above? Then we can use a single `MemoryStream` for both the `jm` and `mj` files.

I'm also a little confused by the way this current is. Does this generate the assembly code _twice_?

Reply:
Updated the code to reuse the `MemoryStream`

Reply:
The assembly code is generated once per typemap file - we have two files to map managed-to-java and java-to-managed.